### PR TITLE
Do not always start a db txn on Postgres

### PR DESCRIPTION
**Pull request by anoadragon453**
While running the unit tests locally and watching my postgres logs, I saw the following log line start to be emitted for every DB query that was being run:

```
2023-01-13 18:05:15.379 GMT [493791] WARNING:  there is already a transaction in progress
```

Which resulted in quite a lot of log lines, generating noise...

This warning is harmless, and is merely telling us that we've started a transaction (with `BEGIN`) and then started another one (with another `BEGIN`). Now this warning will only appear when we actually prepare a database for the first time (which we do all the time when running the unit tests, as each tests uses a separate database) or apply database migrations. 

So while on your actual homeserver you may see this warning pop up a couple times when you update Synapse, it won't be very frequent.

In any case, it should be a harmless change and reduces some noise, mostly for devs :)

### CODE REVIEW

File: synapse/storage/prepare_database.py [hunk 1]
```diff
@@ -108,9 +108,12 @@ def prepare_database(
         # so we start one before running anything. This ensures that any upgrades
         # are either applied completely, or not at all.
         #
-        # (psycopg2 automatically starts a transaction as soon as we run any statements
-        # at all, so this is redundant but harmless there.)
-        cur.execute("BEGIN TRANSACTION")
+        # psycopg2 does automatically start transactions, and while this is technically
```
Comment (DMRobertson) [inline synapse/storage/prepare_database.py:111]: > psycopg2 does automatically start transactions

by _default_, see https://www.psycopg.org/docs/usage.html#transactions-control

File: synapse/storage/prepare_database.py [hunk 2]
```diff
@@ -108,9 +108,12 @@ def prepare_database(
...
+        # harmless to do there as well, doing so results in nested transactions. Those
+        # will generate warnings in Postgres' logs per query, and we'd rather like to
+        # avoid doing that.
+        if isinstance(database_engine, Sqlite3Engine):
```
Comment (DMRobertson) [inline synapse/storage/prepare_database.py:115]: Given the comment above, I think I'd be a little more paranoid: maybe only skip this if we are postgres and the connection does not have autocommit set?

File: synapse/storage/prepare_database.py [hunk 3]
```diff
@@ -108,9 +108,15 @@ def prepare_database(
...
-        # at all, so this is redundant but harmless there.)
-        cur.execute("BEGIN TRANSACTION")
+        # psycopg2 does not automatically start transactions when in autocommit mode.
+        # While it is technically harmless to nest transactions in postgres, doing so
+        # results in a warning in Postgres' logs per query. And we'd rather like to
+        # avoid doing that.
+        if isinstance(database_engine, Sqlite3Engine) or (
+            isinstance(database_engine, PostgresEngine)
+            and db_conn.autocommit
+        ):
```
Comment (clokep) [inline synapse/storage/prepare_database.py:118]: Is `db_conn.autocmmit` always false for sqlite? Can we simplify this to:

```suggestion
        if db_conn.autocommit:
```

Note that attempting to set auto-commit for sqlite seems to be a no-op:

https://github.com/matrix-org/synapse/blob/8c94dd3a277d4e11192f98a9ca32cb6638606b66/synapse/storage/engines/sqlite.py#L114-L119

Comment (anoadragon453) [reply synapse/storage/prepare_database.py:118]: From playing around in the manhole, it sadly appears that `autocommit` isn't a valid attribute in the case of SQLite:

```
>>> hs.get_datastores().main.db_pool.runWithConnection(lambda conn: print("ac:", conn.autocommit))
<Deferred #0>
Deferred #0 failed: "'sqlite3.Connection' object has no attribute 'autocommit'"
```

Comment (DMRobertson) [reply synapse/storage/prepare_database.py:118]: https://docs.python.org/3/library/sqlite3.html?highlight=sqlite3#transaction-control

> The sqlite3 module does not adhere to the transaction handling recommended by [PEP 249](https://peps.python.org/pep-0249/).

🫠🫠

Comment (clokep) [reply synapse/storage/prepare_database.py:118]: That's annoying. I secretly want to burn all `isinstance(foo, FooEngine)`....but not a big deal. 👍

Comment (DMRobertson) [reply synapse/storage/prepare_database.py:118]: It looks like sqlite can be in an autocommit mode or an implicit transaction mode, depending on the value of isolation_level.

It'd be nice if we could abstract these details away into the DB engines somehow. The ideal would be something like 

```
if db_engine.will_autocommit(db_conn):
    cur.execute("BEGIN TRANSACTION")
```

but I don't know if that's going to open up an even bigger can of worms. Maybe see what the rest of the team think?

Comment (squahtx) [reply synapse/storage/prepare_database.py:118]: The abstraction sounds like a good idea. I'd not block the PR on it though.

**Review by squahtx — APPROVED** [review]
lgtm

**Pull request merged**